From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Wed, 12 Jul 2023 19:06:02 +0200
Subject: [PATCH] Add aggressive mob API

Moves the methods from Zombie to Mob, as more Mobs can raise their arm(s)

== AT ==
public net.minecraft.world.entity.Mob DATA_MOB_FLAGS_ID

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index b4e45d8e9851552f30c9c29d3d9671e1e9094c38..ec4ecaaf6ddf6e3b219c9bc34f9f3bb7481846e8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -165,4 +165,16 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         return getHandle().getExperienceReward();
     }
     // Paper end
+
+    // Paper start
+    @Override
+    public boolean isAggressive() {
+        return this.getHandle().isAggressive();
+    }
+
+    @Override
+    public void setAggressive(boolean aggressive) {
+        this.getHandle().setAggressive(aggressive);
+    }
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
index 56f9630dbe5d18d5ec33dc85f6531723022d6a3b..646d915552733603f9884012646d46ed82df5758 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
@@ -127,6 +127,14 @@ public class CraftPanda extends CraftAnimals implements Panda {
         return this.getHandle().getUnhappyCounter();
     }
 
+    // Paper start
+    @Override
+    public boolean isAggressive() {
+        // Pandas are always aggressive if their variant is AGGRESSIVE
+        return (this.getHandle().getEntityData().get(net.minecraft.world.entity.Mob.DATA_MOB_FLAGS_ID) & 4) != 0;
+    }
+    // Paper end
+
     public static Gene fromNms(net.minecraft.world.entity.animal.Panda.Gene gene) {
         Preconditions.checkArgument(gene != null, "Gene may not be null");
 
